service/matching/ack_manager.go

```diff
@@ -38,16 +38,18 @@ import (
 // Used to convert out of order acks into ackLevel movement.
 type ackManager struct {
 	sync.RWMutex
-	outstandingTasks *treemap.Map // TaskID->acked
-	readLevel        int64        // Maximum TaskID inserted into outstandingTasks
-	ackLevel         int64        // Maximum TaskID below which all tasks are acked
+	backlogMgr       *backlogManagerImpl // accessing approximateBacklogCounter
+	outstandingTasks *treemap.Map        // TaskID->acked
+	readLevel        int64               // Maximum TaskID inserted into outstandingTasks
+	ackLevel         int64               // Maximum TaskID below which all tasks are acked
 	backlogCounter   atomic.Int64
 	logger           log.Logger
 }
 
-func newAckManager(logger log.Logger) ackManager {
+func newAckManager(backlogMgr *backlogManagerImpl) ackManager {
```

Comment:
now that I see how the ack manager tests had to get more complex, I kinda wish we did this differently: e.g. accept a function to call on decrement and have backlogManager pass in the right function, and the tests pass in no-op or simple mock. don't have to change it for this pr but it could be nice to clean up later.

the other option is having completeTask _return_ the number of decrements it wants to do and then have backlogManager do the decrement. did we think about that? (again, not suggesting changing the pr at this point, just wondering about alternatives)

Reply:
good point - i shall make a note of it for now and come back to this in the near future